tikv/raftstore/worker.go

```diff
@@ -285,14 +289,118 @@ type pendingDeleteRanges struct {
 	ranges *lockstore.MemStore
 }
 
+func (pendDelRanges *pendingDeleteRanges) insert(regionId uint64, startKey, endKey []byte, timeout time.Time) {
+	// todo, currently it is a place holder.
+}
+
+type delRangeHolder struct {
+	startKey []byte
+	endKey   []byte
+	regionId uint64
+}
+
+// findOverlapRanges finds ranges that overlap with [start_key, end_key).
+func (pendDelRanges *pendingDeleteRanges) findOverlapRanges(startKey, endKey []byte) []delRangeHolder {
+	// todo, currently it is a a place holder.
+	return nil
+}
+
+// drainOverlapRanges gets ranges that overlap with [start_key, end_key).
+func (pendDelRanges *pendingDeleteRanges) drainOverlapRanges(startKey, endKey []byte) []delRangeHolder {
+	ranges := pendDelRanges.findOverlapRanges(startKey, endKey)
+	for _, r := range ranges {
+		y.Assert(pendDelRanges.ranges.Delete(r.startKey))
+	}
+	return ranges
+}
+
 type snapContext struct {
 	engiens             *Engines
-	batchSize           int
+	batchSize           uint64
 	mgr                 *SnapManager
 	cleanStalePeerDelay time.Duration
 	pendingDeleteRanges *pendingDeleteRanges
 }
 
+// handleGen handles the task of generating snapshot of the Region. It calls `generateSnap` to do the actual work.
+func (snapCtx *snapContext) handleGen(regionId uint64, raftSnapshot, kvSnapshot Snapshot, notifier chan<- *eraftpb.Snapshot) {
+	if err := snapCtx.generateSnap(regionId, raftSnapshot, kvSnapshot, notifier); err != nil {
+		log.Errorf("failed to generate snapshot!!!, [regionId: %d, err : %v]", regionId, err)
+	}
+}
+
+// generateSnap generates the snapshots of the Region
+func (snapCtx *snapContext) generateSnap(regionId uint64, raftSnapshot, kvSnapshot Snapshot, notifier chan<- *eraftpb.Snapshot) error {
+	// do we need to check leader here?
+	snap, err := doSnapshot(snapCtx.mgr, raftSnapshot, kvSnapshot, regionId)
+	if err != nil {
+		return err
+	}
+	notifier <- snap
+	return nil
+}
+
+// applySnap applies snapshot data of the Region.
+func (snapCtx *snapContext) applySnap(regionId uint64, status *JobStatus) (int, error) {
+	// todo, currently, it is a place holder.
+	return 0, nil
+}
+
+// handleApply tries to apply the snapshot of the specified Region. It calls `applySnap` to do the actual work.
+func (snapCtx *snapContext) handleApply(regionId uint64, status *JobStatus) {
+	atomic.CompareAndSwapInt64(status, JobStatus_Pending, JobStatus_Running)
+	retCode, err := snapCtx.applySnap(regionId, status)
```

Comment:
We can only return the error, checks it against the predefined errors.

Reply:
ok